### ISSUE TITLE
Better fallbacks for ignored exceptions

### ISSUE THREAD

**Issue by fbeaufume** [resilience4j/resilience4j#804]
Currently working on the migration from Hystrix to Resilience4j 1.2.0 for Spring Boot applications, and facing several issues with "ignoreExceptions".

As a side note, I would like an option to treat some exceptions as a success. I saw in https://github.com/resilience4j/resilience4j/issues/568#issuecomment-521616694 that it could be supported. Any news on that ?

But the main issue here is how Resilience4j uses the fallbacks for ignored exceptions. As described in https://github.com/resilience4j/resilience4j/issues/780#issuecomment-566723745 "ignoreExceptions" does not configure the fallback mecanism. This feels counter-intuitive to me and verbose to address.

For example if all my exceptions extend RuntimeException and I need a fallback for some exceptions (Problem1Exception, Problem2Exception) but not for some other exceptions (Ignored1Exception, Ignored2Exception), then I have to declare two fallbacks (one for each "problem" exception). In addition, I also have to declare an extra fallback for io.github.resilience4j.circuitbreaker.CallNotPermittedException.

I expected to declare a single fallback for RuntimeException, but I actually end up declaring three.

I must either change my exceptions hierarchy to match the Resilience4j behavior or declare multiple fallbacks.

Am I mistaken ? Are there better alternatives ?

**Comment by RobWin** [comment]
Hi,
1) Every exception which is not part of the `recordExceptions` nor `ignoreExceptions` list is recorded as a success. Let's assume you want to handle a `HttpServerErrorException` as a failure, but a `HttpClientErrorException` exception as a success. Just add the `HttpServerErrorException` to the list of `recordExceptions`. The `HttpClientErrorException` is then recorded as a success.

2) Your expectation would be that ignored exceptions are not handled by a generic fallback method? If more and more people would expect this behavior (because of Hystrix), we might need to change it. @Romeh What do you think? But this change would not be backwards compatible. 

I think currently the only alternative is to use `instanceof` checks in your fallback method which catches `RuntimeExecption`.

**Comment by artemzi** [comment]
@RobWin +1 for the possibility to mark ignnoredExceptions as successful or failed responses. The main idea here to see these requests in monitoring dashboards. I don't want to open circuit in some scenarios of service requests exceptions, but also don't want these requests to be completely ignored in statistics

**Comment by RobWin** [comment]
As I said. Just configure `recordExceptions` and every other exception is counted as a success and is shown in your metrics. There is also a metric for ignored exceptions. They are not completly ignored in statistics.

Please see https://github.com/resilience4j/resilience4j/blob/master/resilience4j-micrometer/src/main/java/io/github/resilience4j/micrometer/tagged/AbstractCircuitBreakerMetrics.java#L118

**Comment by artemzi** [comment]
Thanks, I just start to look around resilience4j and looking for a possibility for migration from hystrix. I use `ignoreExceptions` for preventing open circuits in a specific case.

**Comment by fbeaufume** [comment]
Thank you @RobWin for the feedback.

Unfortunately, "Just configure recordExceptions and every other exception is recorded as a success" means that exceptions that I did not anticipate will be treated as success. This "success by default" model does not feel safe, or resilient ;), enough to me.

In addition I still have to declare a fallback for each failure exception + one fallback for `io.github.resilience4j.circuitbreaker.CallNotPermittedException` (or else the callers will receive that exception when the CB is open).

My initial strategy was (using a simplified syntax):

- "`resilience4j.circuitbreaker.instances.myconfig.ignoreExceptions: Success1Exception, Success2Exception`" in the config
- `String myFallback(Failure1Exception ex) { ... }`
- `String myFallback(Failure2Exception ex) { ... }`
- `String myFallback(CallNotPermittedException ex) { ... }`

My understading is that your suggestion is:

- "`resilience4j.circuitbreaker.instances.myconfig.recordExceptions: Failure1Exception, Failure2Exception`" in the config
- `String myFallback(Failure1Exception ex) { ... }`
- `String myFallback(Failure2Exception ex) { ... }`
- `String myFallback(CallNotPermittedException ex) { ... }`

Is there a simpler approach ?

I must admit that I'm more confortable with the Hystrix model where the circuit breaker state and the fallback call are correlated:

- `@HystrixCommand(..., ignoreExceptions = {Success1Exception.class,Success2Exception.class})` on the business method
- `String myFallback(Throwable t) { ... }` (or just `String myFallback() { ... }`)

**Comment by RobWin** [comment]
Actually from my experience it's more resilient like it is today. Just imagine your code or some library which you use throws runtime exceptions which you did not anticipate but are thrown before you call the remote service, you don't want to open the CircuitBreaker in that case for all calls. Usually you better know upfront which kind of exceptions signal to you that the remote service is unavailable.

**Comment by fbeaufume** [comment]
So, in my previous comment, is my understanding of your suggestion correct ?
Is there a simpler approach ?

**Comment by RobWin** [comment]
Yes, your understanding is correct. Still any improvement suggestions are welcome. 
As far as I know Hystrix does not allow to record exceptions as a success. They don't have the problem to intersect 3 lists. 
We could introduce a `successExceptions` list, but the intersection is not simple.

**Comment by RobWin** [comment]
You can still just use one fallback method and implement a dispatcher based on instanceof.

**Comment by RobWin** [comment]
@Romeh but I must agree that its confusing that ignore exceptions are handled by the fallback mechanism.

**Comment by egch** [comment]
> We could introduce a `successExceptions` list, but the intersection is not simple.

Was this feature, successExceptions, eventually implemented in any newer version?

**Comment by RobWin** [comment]
No
